File: app/src/main/java/com/veyndan/paper/reddit/post/delegate/PostAdapterDelegate.java
package com.veyndan.paper.reddit.post.delegate;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.customtabs.CustomTabsClient;
import android.support.customtabs.CustomTabsIntent;
import android.support.customtabs.CustomTabsServiceConnection;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;

import com.hannesdorfmann.adapterdelegates3.AdapterDelegate;
import com.jakewharton.rxbinding2.support.design.widget.RxSnackbar;
import com.jakewharton.rxbinding2.view.RxView;
import com.jakewharton.rxbinding2.widget.RxCompoundButton;
import com.jakewharton.rxbinding2.widget.RxPopupMenu;
import com.veyndan.paper.reddit.MainActivity;
import com.veyndan.paper.reddit.R;
import com.veyndan.paper.reddit.api.reddit.Reddit;
import com.veyndan.paper.reddit.api.reddit.model.Listing;
import com.veyndan.paper.reddit.api.reddit.model.Thing;
import com.veyndan.paper.reddit.api.reddit.network.VoteDirection;
import com.veyndan.paper.reddit.databinding.PostItemBinding;
import com.veyndan.paper.reddit.post.Flair;
import com.veyndan.paper.reddit.post.PostAdapter;
import com.veyndan.paper.reddit.post.media.PostMediaAdapter;
import com.veyndan.paper.reddit.post.model.Post;
import com.veyndan.paper.reddit.ui.recyclerview.Swipeable;
import com.veyndan.paper.reddit.util.Node;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;
import timber.log.Timber;

public class PostAdapterDelegate extends AdapterDelegate<List<Node<Response<Thing<Listing>>>>> {

    private static final String CUSTOM_TAB_PACKAGE_NAME = "com.android.chrome";

    private final PostAdapter adapter;
    private final Activity activity;
    private final Reddit reddit;

    private final CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
    private final CustomTabsIntent customTabsIntent = builder.build();
    @Nullable private CustomTabsClient customTabsClient;

    public PostAdapterDelegate(final PostAdapter adapter, final Activity activity, final Reddit reddit) {
        this.adapter = adapter;
        this.activity = activity;
        this.reddit = reddit;

        CustomTabsClient.bindCustomTabsService(activity, CUSTOM_TAB_PACKAGE_NAME, new CustomTabsServiceConnection() {
            @Override
            public void onCustomTabsServiceConnected(final ComponentName name, final CustomTabsClient client) {
                // customTabsClient is now valid.
                customTabsClient = client;
                customTabsClient.warmup(0);
            }

            @Override
            public void onServiceDisconnected(final ComponentName name) {
                // customTabsClient is no longer valid. This also invalidates sessions.
                customTabsClient = null;
            }
        });
    }

    @Override
    public boolean isForViewType(@NonNull final List<Node<Response<Thing<Listing>>>> nodes,
                                 final int position) {
        return nodes.get(position) instanceof Post;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(final ViewGroup parent) {
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        final PostItemBinding binding = PostItemBinding.inflate(inflater, parent, false);
        return new PostViewHolder(binding, adapter, reddit);
    }

    @Override
    protected void onBindViewHolder(@NonNull final List<Node<Response<Thing<Listing>>>> nodes,
                                    final int position,
                                    @NonNull final RecyclerView.ViewHolder holder,
                                    @NonNull final List<Object> payloads) {
        final Context context = holder.itemView.getContext();
        final PostViewHolder postHolder = (PostViewHolder) holder;
        final Post post = (Post) nodes.get(position);

        bindHeader(context, post, postHolder);
        bindMedia(post, postHolder, activity, customTabsClient, customTabsIntent);
        bindPoints(context, post, postHolder);
        bindUpvoteAction(context, post, postHolder, reddit);
        bindDownvoteAction(context, post, postHolder, reddit);
        bindSaveAction(post, postHolder, reddit);
        bindCommentsAction(context, nodes, post, postHolder, adapter);

        final PopupMenu otherMenu = new PopupMenu(context, postHolder.binding.postOther);
        otherMenu.getMenuInflater().inflate(R.menu.menu_post_other, otherMenu.getMenu());

        RxView.clicks(postHolder.binding.postOther)
                .subscribe(aVoid -> otherMenu.show(), Timber::e);

        RxPopupMenu.itemClicks(otherMenu)
                .subscribe(menuItem -> {
                    switch (menuItem.getItemId()) {
                        case R.id.action_post_share:
                            bindShareAction(context, post);
                            break;
                        case R.id.action_post_browser:
                            bindBrowserAction(context, post);
                            break;
                        case R.id.action_post_report:
                            break;
                    }
                }, Timber::e);
    }

    private static void bindHeader(final Context context, final Post post, final PostViewHolder holder) {
        final List<Flair> flairs = new ArrayList<>();

        if (post.stickied()) {
            final int flairStickiedColor = context.getColor(R.color.post_flair_stickied);
            final String flairStickiedText = context.getString(R.string.post_stickied);

            flairs.add(Flair.builder(flairStickiedColor)
                    .text(flairStickiedText)
                    .build());
        }

        if (post.locked()) {
            final int flairLockedColor = context.getColor(R.color.post_flair_locked);
            final String flairLockedText = context.getString(R.string.post_locked);
            final Drawable flairLockIcon = context.getDrawable(R.drawable.ic_lock_outline_white_12sp);

            flairs.add(Flair.builder(flairLockedColor)
                    .text(flairLockedText)
                    .icon(flairLockIcon)
                    .build());
        }

        if (post.nsfw()) {
            final int flairNsfwColor = context.getColor(R.color.post_flair_nsfw);
            final String flairNsfwText = context.getString(R.string.post_nsfw);

            flairs.add(Flair.builder(flairNsfwColor)
                    .text(flairNsfwText)
                    .build());
        }

        if (post.hasLinkFlair()) {
            final int flairLinkColor = context.getColor(R.color.post_flair_link);

            flairs.add(Flair.builder(flairLinkColor)
                    .type(Flair.Type.LINK)
                    .text(post.linkFlair())
                    .build());
        }

        if (post.isGilded()) {
            final int flairGildedColor = context.getColor(R.color.post_flair_gilded);
            final Drawable flairGildedIcon = context.getDrawable(R.drawable.ic_star_white_12sp);

            flairs.add(Flair.builder(flairGildedColor)
                    .text(String.valueOf(post.gildedCount()))
                    .icon(flairGildedIcon)
                    .build());
        }

        holder.binding.postHeader.setHeader(post.linkTitle(), post.author(), post.getDisplayAge(),
                post.subreddit(), flairs);
    }

    private static void bindMedia(final Post post, final PostViewHolder holder,
                                  final Activity activity, final CustomTabsClient customTabsClient,
                                  final CustomTabsIntent customTabsIntent) {
        post.medias().value
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .toList()
                .subscribe(medias -> {
                    final PostMediaAdapter postMediaAdapter = new PostMediaAdapter(
                            activity, customTabsClient, customTabsIntent, post, medias);
                    holder.binding.postMediaView.setAdapter(postMediaAdapter);
                });
    }

    private static void bindPoints(final Context context, final Post post, final PostViewHolder holder) {
        final String points = post.getDisplayPoints(context);
        holder.binding.postScore.setText(points);
    }

    private static void bindUpvoteAction(final Context context, final Post post,
                                         final PostViewHolder holder, final Reddit reddit) {
        final VoteDirection likes = post.likes().value;
        holder.binding.postUpvoteNew.setChecked(likes == VoteDirection.UPVOTE);
        RxCompoundButton.checkedChanges(holder.binding.postUpvoteNew)
                // checkedChanges emits the checked state on subscription. As the voted state of
                // the Reddit post is the same as the checked state of the button initially,
                // skipping the initial emission means no unnecessary network requests occur.
                .skip(1)
                .filter(isChecked -> !post.archived())
                .subscribe(isChecked -> {
                    post.likes().value = isChecked ? VoteDirection.UPVOTE : VoteDirection.UNVOTE;
                    reddit.vote(isChecked ? VoteDirection.UPVOTE : VoteDirection.UNVOTE, post.fullname())
                            .subscribeOn(Schedulers.io())
                            .subscribe(response -> {}, Timber::e);

                    post.points().value += isChecked ? 1 : -1;

                    final String points1 = post.getDisplayPoints(context);
                    holder.binding.postScore.setText(points1);
                }, Timber::e);
    }

    private static void bindDownvoteAction(final Context context, final Post post,
                                           final PostViewHolder holder, final Reddit reddit) {
        final VoteDirection likes = post.likes().value;
        holder.binding.postDownvoteNew.setChecked(likes == VoteDirection.DOWNVOTE);
        RxCompoundButton.checkedChanges(holder.binding.postDownvoteNew)
                // checkedChanges emits the checked state on subscription. As the voted state of
                // the Reddit post is the same as the checked state of the button initially,
                // skipping the initial emission means no unnecessary network requests occur.
                .skip(1)
                .filter(isChecked -> !post.archived())
                .subscribe(isChecked -> {
                    post.likes().value = isChecked ? VoteDirection.DOWNVOTE : VoteDirection.UNVOTE;
                    reddit.vote(isChecked ? VoteDirection.DOWNVOTE : VoteDirection.UNVOTE, post.fullname())
                            .subscribeOn(Schedulers.io())
                            .subscribe(response -> {}, Timber::e);

                    post.points().value += isChecked ? -1 : 1;

                    final String points1 = post.getDisplayPoints(context);
                    holder.binding.postScore.setText(points1);
                }, Timber::e);
    }

    private static void bindSaveAction(final Post post, final PostViewHolder holder,
                                       final Reddit reddit) {
        holder.binding.postSave.setChecked(post.saved().value);
        RxView.clicks(holder.binding.postSave)
                .subscribe(aVoid -> {
                    holder.binding.postSave.toggle();
                    final boolean isChecked = holder.binding.postSave.isChecked();

                    post.saved().value = isChecked;
                    if (isChecked) {
                        reddit.save("", post.fullname())
                                .subscribeOn(Schedulers.io())
                                .subscribe(response -> {}, Timber::e);
                    } else {
                        reddit.unsave(post.fullname())
                                .subscribeOn(Schedulers.io())
                                .subscribe(response -> {}, Timber::e);
                    }
                }, Timber::e);
    }

    private static void bindCommentsAction(final Context context,
                                           final List<Node<Response<Thing<Listing>>>> nodes,
                                           final Post post, final PostViewHolder holder,
                                           final PostAdapter adapter) {
        RxView.clicks(holder.binding.postComments)
                .map(aVoid -> {
                    holder.binding.postComments.toggle();
                    return holder.binding.postComments.isChecked();
                })
                .subscribe(displayDescendants -> {
                    post.descendantsVisible().value = !displayDescendants;
                    if (post.comment()) {
                        if (displayDescendants) {
                            int i;
                            for (i = holder.getAdapterPosition() + 1; i < nodes.size() && nodes.get(i).depth() > post.depth(); i++)
                                ;

                            nodes.subList(holder.getAdapterPosition() + 1, i).clear();
                            adapter.notifyItemRangeRemoved(holder.getAdapterPosition() + 1, i - (holder.getAdapterPosition() + 1));

                            holder.binding.postCommentCount.setVisibility(View.VISIBLE);
                            holder.binding.postCommentCount.setText(String.valueOf(i - (holder.getAdapterPosition() + 1)));
                        } else {
                            post.preOrderTraverse(post.depth())
                                    .skip(1)
                                    .toList()
                                    .subscribe(children -> {
                                        nodes.addAll(holder.getAdapterPosition() + 1, children);
                                        adapter.notifyItemRangeInserted(holder.getAdapterPosition() + 1, children.size());
                                    }, Timber::e);

                            holder.binding.postCommentCount.setVisibility(View.INVISIBLE);
                        }
                    } else {
                        final Intent commentsIntent = new Intent(context, MainActivity.class);
                        commentsIntent.putExtras(new Reddit.FilterBuilder()
                                .nodeDepth(0)
                                .commentsSubreddit(post.subreddit())
                                .commentsArticle(post.article())
                                .build());
                        context.startActivity(commentsIntent);
                    }
                }, Timber::e);

        if (post.internalNode() && !post.descendantsVisible().value) {
            holder.binding.postCommentCount.setVisibility(View.VISIBLE);
            final String commentCount = post.getDisplayDescendants();
            holder.binding.postCommentCount.setText(commentCount);
        } else {
            holder.binding.postCommentCount.setVisibility(View.INVISIBLE);
        }
    }

    private static void bindShareAction(final Context context, final Post post) {
        final Intent intent = new Intent(Intent.ACTION_SEND);
        intent.putExtra(Intent.EXTRA_TEXT, post.permalink());
        intent.setType("text/plain");
        context.startActivity(intent);
    }

    private static void bindBrowserAction(final Context context, final Post post) {
        final Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(post.linkUrl()));
        context.startActivity(intent);
    }

    static class PostViewHolder extends RecyclerView.ViewHolder implements Swipeable {

        private final PostItemBinding binding;
        private final PostAdapter adapter;
        private final Reddit reddit;

        PostViewHolder(final PostItemBinding binding, final PostAdapter adapter, final Reddit reddit) {
            super(binding.getRoot());

            this.binding = binding;
            this.adapter = adapter;
            this.reddit = reddit;
        }

        @Override
        public boolean swipeable() {
            final int position = getAdapterPosition();
            final Post post = (Post) adapter.getItems().get(position);
            return post.hideable();
        }

        @Override
        public void onSwipe() {
            final int position = getAdapterPosition();
            final Node<Response<Thing<Listing>>> node = adapter.getItems().get(position);
            final Post post = (Post) node;

            final View.OnClickListener undoClickListener = view -> {
                // If undo pressed, then don't follow through with request to hide
                // the post.
                adapter.getItems().add(position, node);
                adapter.notifyItemInserted(position);
            };

            final Snackbar snackbar = Snackbar.make(itemView, R.string.notify_post_hidden, Snackbar.LENGTH_LONG)
                    .setAction(R.string.notify_post_hidden_undo, undoClickListener);

            RxSnackbar.dismisses(snackbar)
                    // If undo pressed, don't hide post.
                    .filter(event -> event != Snackbar.Callback.DISMISS_EVENT_ACTION)
                    .firstElement()
                    .subscribe(event -> {
                        // Chance to undo post hiding has gone, so follow through with
                        // hiding network request.
                        reddit.hide(post.fullname())
                                .subscribeOn(Schedulers.io())
                                .subscribe(response -> {}, Timber::e);
                    }, Timber::e);

            snackbar.show();

            // Hide post from list, but make no network request yet. Outcome of the
            // user's interaction with the snackbar handling will determine this.
            adapter.getItems().remove(position);
            adapter.notifyItemRemoved(position);
        }
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/post/delegate/PostAdapterDelegate.kt
package com.veyndan.paper.reddit.post.delegate

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.support.customtabs.CustomTabsClient
import android.support.customtabs.CustomTabsIntent
import android.support.customtabs.CustomTabsServiceConnection
import android.support.design.widget.Snackbar
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import com.hannesdorfmann.adapterdelegates3.AdapterDelegate
import com.jakewharton.rxbinding2.support.design.widget.dismisses
import com.jakewharton.rxbinding2.view.clicks
import com.jakewharton.rxbinding2.widget.checkedChanges
import com.jakewharton.rxbinding2.widget.itemClicks
import com.veyndan.paper.reddit.MainActivity
import com.veyndan.paper.reddit.R
import com.veyndan.paper.reddit.api.reddit.Reddit
import com.veyndan.paper.reddit.api.reddit.model.Listing
import com.veyndan.paper.reddit.api.reddit.model.Thing
import com.veyndan.paper.reddit.api.reddit.network.VoteDirection
import com.veyndan.paper.reddit.databinding.PostItemBinding
import com.veyndan.paper.reddit.post.Flair
import com.veyndan.paper.reddit.post.PostAdapter
import com.veyndan.paper.reddit.post.media.PostMediaAdapter
import com.veyndan.paper.reddit.post.model.Post
import com.veyndan.paper.reddit.ui.recyclerview.Swipeable
import com.veyndan.paper.reddit.util.Node
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import retrofit2.Response

class PostAdapterDelegate(private val adapter: PostAdapter, private val activity: Activity,
                          private val reddit: Reddit)
    : AdapterDelegate<MutableList<Node<Response<Thing<Listing>>>>>() {

    companion object {

        private const val CUSTOM_TAB_PACKAGE_NAME: String = "com.android.chrome"
    }

    private val builder: CustomTabsIntent.Builder = CustomTabsIntent.Builder()
    private val customTabsIntent: CustomTabsIntent = builder.build()
    private var customTabsClient: CustomTabsClient? = null

    init {
        CustomTabsClient.bindCustomTabsService(activity, CUSTOM_TAB_PACKAGE_NAME, object : CustomTabsServiceConnection() {
            override fun onCustomTabsServiceConnected(name: ComponentName?, client: CustomTabsClient?) {
                // customTabsClient is now valid.
                customTabsClient = client
                customTabsClient!!.warmup(0)
            }

            override fun onServiceDisconnected(name: ComponentName?) {
                // customTabsClient is no longer valid. This also invalidates sessions.
                customTabsClient = null
            }
        })
    }

    override fun isForViewType(nodes: MutableList<Node<Response<Thing<Listing>>>>, position: Int): Boolean = nodes[position] is Post

    override fun onCreateViewHolder(parent: ViewGroup?): RecyclerView.ViewHolder {
        val inflater: LayoutInflater = LayoutInflater.from(parent!!.context)
        val binding: PostItemBinding = PostItemBinding.inflate(inflater, parent, false)
        return PostViewHolder(binding, adapter, reddit)
    }

    override fun onBindViewHolder(nodes: MutableList<Node<Response<Thing<Listing>>>>, position: Int, holder: RecyclerView.ViewHolder, payloads: List<Any>) {
        val context: Context = holder.itemView.context
        val postHolder: PostViewHolder = holder as PostViewHolder
        val post: Post = nodes[position] as Post

        if (payloads.isEmpty()) {
            bindTitle(post, postHolder)
            bindSubtitle(post, postHolder)
            bindFlairs(context, post, postHolder)
            bindMedia(post, postHolder, activity, customTabsClient, customTabsIntent)
            bindPoints(context, post, postHolder)
            bindActions(context, nodes, post, postHolder, reddit, adapter)
        } else if (payloads[0] == PostPayload.VOTE) {
            bindPoints(context, post, postHolder)
        }
    }

    private fun bindTitle(post: Post, holder: PostViewHolder) {
        holder.binding.postTitle.setTitle(post.linkTitle!!)
    }

    private fun bindSubtitle(post: Post, holder: PostViewHolder) {
        holder.binding.postSubtitle.setSubtitle(post.author, post.getDisplayAge(), post.subreddit)
    }

    private fun bindFlairs(context: Context, post: Post, holder: PostViewHolder) {
        val flairs: Collection <Flair> = post.flairs(context)
        holder.binding.postFlairs.setFlairs(flairs, post.subreddit)
    }

    private fun bindMedia(post: Post, holder: PostViewHolder, activity: Activity,
                          customTabsClient: CustomTabsClient?, customTabsIntent: CustomTabsIntent) {
        post.medias
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .toList()
                .subscribe { medias ->
                    val postMediaAdapter: PostMediaAdapter = PostMediaAdapter(
                            activity, customTabsClient, customTabsIntent, post, medias)
                    holder.binding.postMediaView.adapter = postMediaAdapter
                }
    }

    private fun bindPoints(context: Context, post: Post, holder: PostViewHolder) {
        val points: String = post.getDisplayPoints(context)
        holder.binding.postScore.text = points
    }

    private fun bindActions(context: Context, nodes: MutableList<Node<Response<Thing<Listing>>>>,
                            post: Post, holder: PostViewHolder, reddit: Reddit,
                            adapter: PostAdapter) {
        bindUpvoteAction(post, holder, reddit, adapter)
        bindDownvoteAction(post, holder, reddit, adapter)
        bindSaveAction(post, holder, reddit)
        bindCommentsAction(context, nodes, post, holder, adapter)
        bindPopupActions(context, post, holder)
    }

    private fun bindUpvoteAction(post: Post, holder: PostViewHolder, reddit: Reddit,
                                 adapter: PostAdapter) {
        val likes: VoteDirection = post.likes
        holder.binding.postUpvoteNew.isChecked = likes == VoteDirection.UPVOTE
        holder.binding.postUpvoteNew.checkedChanges()
                // checkedChanges emits the checked state on subscription. As the voted state of
                // the Reddit post is the same as the checked state of the button initially,
                // skipping the initial emission means no unnecessary network requests occur.
                .skip(1)
                .filter { !post.archived }
                .subscribe { isChecked ->
                    post.likes = if (isChecked) VoteDirection.UPVOTE else VoteDirection.UNVOTE
                    reddit.vote(post.likes, post.fullname)
                            .subscribeOn(Schedulers.io())
                            .subscribe()

                    post.points += if (isChecked) 1 else -1

                    adapter.notifyItemChanged(holder.adapterPosition, PostPayload.VOTE)
                }
    }

    private fun bindDownvoteAction(post: Post, holder: PostViewHolder, reddit: Reddit,
                                   adapter: PostAdapter) {
        val likes: VoteDirection = post.likes
        holder.binding.postDownvoteNew.isChecked = likes == VoteDirection.DOWNVOTE
        holder.binding.postDownvoteNew.checkedChanges()
                // checkedChanges emits the checked state on subscription. As the voted state of
                // the Reddit post is the same as the checked state of the button initially,
                // skipping the initial emission means no unnecessary network requests occur.
                .skip(1)
                .filter { !post.archived }
                .subscribe { isChecked ->
                    post.likes = if (isChecked) VoteDirection.DOWNVOTE else VoteDirection.UNVOTE
                    reddit.vote(post.likes, post.fullname)
                            .subscribeOn(Schedulers.io())
                            .subscribe()

                    post.points += if (isChecked) -1 else 1

                    adapter.notifyItemChanged(holder.adapterPosition, PostPayload.VOTE)
                }
    }

    private fun bindSaveAction(post: Post, holder: PostViewHolder, reddit: Reddit) {
        holder.binding.postSave.isChecked = post.saved
        holder.binding.postSave.clicks()
                .subscribe {
                    holder.binding.postSave.toggle()
                    val isChecked: Boolean = holder.binding.postSave.isChecked

                    post.saved = isChecked
                    if (isChecked) {
                        reddit.save("", post.fullname)
                                .subscribeOn(Schedulers.io())
                                .subscribe()
                    } else {
                        reddit.unsave(post.fullname)
                                .subscribeOn(Schedulers.io())
                                .subscribe()
                    }
                }
    }

    private fun bindCommentsAction(context: Context, nodes: MutableList<Node<Response<Thing<Listing>>>>, post: Post, holder: PostViewHolder, adapter: PostAdapter) {
        holder.binding.postComments.clicks()
                .map {
                    holder.binding.postComments.toggle()
                    holder.binding.postComments.isChecked
                }
                .subscribe { displayDescendants ->
                    post.descendantsVisible = !displayDescendants
                    if (post.comment) {
                        if (displayDescendants) {
                            var i: Int = holder.adapterPosition + 1
                            while (i < nodes.size && nodes[i].depth > post.depth) {
                                i++
                            }

                            nodes.subList(holder.adapterPosition + 1, i).clear()
                            adapter.notifyItemRangeRemoved(holder.adapterPosition + 1, i - (holder.adapterPosition + 1))

                            holder.binding.postCommentCount.visibility = View.VISIBLE
                            holder.binding.postCommentCount.text = (i - (holder.adapterPosition + 1)).toString()
                        } else {
                            post.preOrderTraverse(post.depth)
                                    .skip(1)
                                    .toList()
                                    .subscribe { children ->
                                        nodes.addAll(holder.adapterPosition + 1, children)
                                        adapter.notifyItemRangeInserted(holder.adapterPosition + 1, children.size)
                                    }

                            holder.binding.postCommentCount.visibility = View.INVISIBLE
                        }
                    } else {
                        val commentsIntent: Intent = Intent(context, MainActivity::class.java)
                        commentsIntent.putExtra(Reddit.FILTER, Reddit.Filter(
                                nodeDepth = 0,
                                commentsSubreddit = post.subreddit,
                                commentsArticle = post.article()))
                        context.startActivity(commentsIntent)
                    }
                }

        post.internalNode()
                .subscribe { internalNode ->
                    if (internalNode && !post.descendantsVisible) {
                        holder.binding.postCommentCount.visibility = View.VISIBLE
                        post.getDisplayDescendants()
                                .subscribe { descendants -> holder.binding.postCommentCount.text = descendants }
                    } else {
                        holder.binding.postCommentCount.visibility = View.INVISIBLE
                    }
                }
    }

    private fun bindPopupActions(context: Context, post: Post, holder: PostViewHolder) {
        val otherMenu: PopupMenu = PopupMenu(context, holder.binding.postOther)
        otherMenu.menuInflater.inflate(R.menu.menu_post_other, otherMenu.menu)

        holder.binding.postOther.clicks()
                .subscribe { otherMenu.show() }

        otherMenu.itemClicks()
                .subscribe { menuItem ->
                    when (menuItem.itemId) {
                        R.id.action_post_share -> bindShareAction(context, post)
                        R.id.action_post_browser -> bindBrowserAction(context, post)
                        R.id.action_post_report -> {
                        }
                    }
                }
    }

    private fun bindShareAction(context: Context, post: Post) {
        val intent: Intent = Intent(Intent.ACTION_SEND)
        intent.putExtra(Intent.EXTRA_TEXT, post.permalink)
        intent.type = "text/plain"
        context.startActivity(intent)
    }

    private fun bindBrowserAction(context: Context, post: Post) {
        val intent: Intent = Intent(Intent.ACTION_VIEW, Uri.parse(post.linkUrl))
        context.startActivity(intent)
    }

    private enum class PostPayload {
        VOTE
    }

    class PostViewHolder(val binding: PostItemBinding, private val adapter: PostAdapter,
                         private val reddit: Reddit)
        : RecyclerView.ViewHolder(binding.root), Swipeable {

        override fun swipeable(): Boolean {
            val post: Post = adapter.items[adapterPosition] as Post
            return post.hideable
        }

        override fun onSwipe() {
            val position: Int = adapterPosition
            val node: Node <Response <Thing <Listing>>> = adapter.items[position]
            val post: Post = node as Post

            val undoClickListener: View.OnClickListener = View.OnClickListener {
                // If undo pressed, then don't follow through with request to hide
                // the post.
                adapter.items.add(position, node)
                adapter.notifyItemInserted(position)
            }

            val snackbar: Snackbar = Snackbar.make(itemView, R.string.notify_post_hidden, Snackbar.LENGTH_LONG)
                    .setAction(R.string.notify_post_hidden_undo, undoClickListener)

            snackbar.dismisses()
                    // If undo pressed, don't hide post.
                    .filter { event -> event != Snackbar.Callback.DISMISS_EVENT_ACTION }
                    .firstElement()
                    .subscribe {
                        // Chance to undo post hiding has gone, so follow through with
                        // hiding network request.
                        reddit.hide(post.fullname)
                                .subscribeOn(Schedulers.io())
                                .subscribe()
                    }

            snackbar.show()

            // Hide post from list, but make no network request yet. Outcome of the
            // user's interaction with the snackbar handling will determine this.
            adapter.items.removeAt(position)
            adapter.notifyItemRemoved(position)
        }
    }
}
